File: src/main/java/ar/edu/unlam/tallerweb1/infrastructure/JdbcRepository.java
package ar.edu.unlam.tallerweb1.infrastructure;

import ar.edu.unlam.tallerweb1.infrastructure.utils.DataSource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.Connection;
import java.sql.SQLException;
import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Consumer;

public abstract class JdbcRepository {

    private DataSource dataSource;

    public JdbcRepository(DataSource dataSource){

        this.dataSource = dataSource;
    }
    private static Logger logger = LoggerFactory.getLogger(JdbcRepository.class);
    protected <T> List<T> searchInTransaction(BiFunction<Connection, String, List<T>> function, String inputValue){
        try (final Connection conn = dataSource.getConnection()){
            return function.apply(conn, inputValue);

        } catch (SQLException e) {
            error(e);
            throw new RuntimeException(e);
        }
    }

    protected <T> T findInTransaction(BiFunction<Connection, String, T> function, String inputValue){
        try (final Connection conn = dataSource.getConnection()){
            return function.apply(conn, inputValue);
        } catch (SQLException e) {
            error(e);
            throw new RuntimeException(e);
        }
    }

    protected <T> void executeInTransaction(Consumer<Connection> function){
        try (final Connection conn = dataSource.getConnection()){
            function.accept(conn);

        } catch (SQLException e) {
            error(e);
            throw new RuntimeException(e);
        }
    }

    private static void error(Exception e) {
        logger.error(e.getMessage(), e);
    }
}

File: src/main/java/ar/edu/unlam/tallerweb1/infrastructure/JdbcRepository.kt
package ar.edu.unlam.tallerweb1.infrastructure

import ar.edu.unlam.tallerweb1.infrastructure.utils.DataSource
import org.slf4j.Logger
import org.slf4j.LoggerFactory.getLogger
import java.sql.Connection
import java.sql.SQLException

abstract class JdbcRepository(private val dataSource: DataSource) {
    private val logger: Logger = getLogger(JdbcRepository::class.java)

    @Throws(RuntimeException::class)
    protected fun <T> searchInTransaction(function: (Connection, String) -> List<T>, inputValue: String):  List<T> {
        try{
            return function(dataSource.getConnection(), inputValue)
        } catch(e: SQLException) {
            logger.error(e.message, e)
            throw RuntimeException(e);
        }
    }

    protected fun <T> findInTransaction(function: (Connection, String) -> T?, inputValue: String): T?{
        try{
            return function(dataSource.getConnection(), inputValue)
        } catch(e: SQLException) {
            logger.error(e.message, e)
            throw RuntimeException(e);
        }
    }

    protected fun executeInTransaction(function: (Connection) -> Unit) {
        try{
            function(dataSource.getConnection())
        } catch(e: SQLException) {
            logger.error(e.message, e)
            throw RuntimeException(e);
        }
    }

}